### ISSUE TITLE
Subgroup Ballots and Remainder Subgroups

### ISSUE THREAD

**Issue by bashbaug** [KhronosGroup/OpenCL-Docs#626]
The extended subgroup ballot functions `sub_group_ballot_bit_count`, `sub_group_ballot_find_lsb`, and `sub_group_ballot_find_msb` are currently written to only consider bits in the ballot corresponding to work-items in the subgroup.  Example for `sub_group_ballot_bit_count`:

> Returns the number of bits that are set in the bitfield value, only considering the bits in value that represent predicate values from all work items in the subgroup.

The size of the subgroup isn't known at compile-time though, so an implementation may need special code to consider the correct number of bits in the bitfield at runtime.  For some implementations, it would be more efficient to consider the **maximum** subgroup size instead, which is known at compile-time.

**Are these functions intended to work on the maximum subgroup size instead (possibly improving efficiency), or is it important that they only work on the work-items in the subgroup?**

Note that for most typical usages that operate on ballots obtained via subgroup built-ins there is no difference in functionality between the two.  This is because a subgroup ballot bit is set if and only if a) the work-item is active, and b) the predicate is true.  Therefore, the ballot bits corresponding to work-items in "full" subgroups but not in the "remainder" subgroup are unset and hence do not modify the bit count nor affect the least or most significant bit set:

```c
// This will return the same count regardless whether the mask bits
// corresponding to work-items in a "partial" subgroup are considered or
// whether mask bits for "full" subgroups are considred unconditionally.
uint4 my_ballot = sub_group_ballot(my_condition);
uint my_count = sub_group_ballot_bit_count(my_ballot);
```

If the ballot is manipulated manually, the results can be different.  Consider, for example:

```c
uint4 my_manipulated_ballot = ~0; // all bits set

// Is this the subgroup size or the max subgroup size?
uint my_count = sub_group_ballot_bit_count(my_ballot);
```

This doesn't seem to be a common case though.

A few other brief notes worth mentioning:

* I initially thought that the ballot scans could be affected also, but since they only consider bits up to this work-item's subgroup local ID I don't think they are affected.
* Vulkan has a similar issue in theory, although subgroups in Vulkan are a little different, and a remainder subgroup in Vulkan has the same subgroup size as a full subgroup it's just that some of the work-items in the subgroup are inactive by default ([link](https://www.khronos.org/blog/vulkan-subgroup-tutorial)).
* The SPIR-V spec just says "considering only the bits in Value required to represent all bits of the group’s invocations", which could be interpreted either way ([example link](https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#OpGroupNonUniformBallotBitCount)).

**Comment by bashbaug** [comment]
Minor update: `sub_group_ballot_find_lsb` is unaffected as currently written because the spec says:

> If no bits representing predicate values from all work items in the subgroup are set in the bitfield value then the return value is undefined.

So, I believe this question only apples to `sub_group_ballot_bit_count` and `sub_group_ballot_find_msb`.

**Comment by StuartDBrady** [comment]
@bashbaug I believe in our discussion of this previously, we concluded that the spec was intended to refer to the maximum subgroup size, and not the subgroup size excluding non-existent work-items in the "remainder" subgroup.

I agree that `sub_group_ballot_find_lsb` is unaffected.

Are we just awaiting a PR to the spec to clarify this, now?  If so, I could perhaps look to see how this could be worded.

**Comment by bashbaug** [comment]
> Are we just awaiting a PR to the spec to clarify this, now? If so, I could perhaps look to see how this could be worded.

I think so.  It's been a while!  If you're able to take a look at this I think it would help to move things forward.  Thanks!

**Comment by bashbaug** [comment]
This was fixed by #689.  Closing!
